### Project Name : GDS Mart's - Supply Chain Efficiency Analysis

#### Problem statement : GDS Mart, an FMCG manufacturer in Gujarat, India, aims to expand to new cities but faces customer retention issues due to delivery service problems. They seek to measure 'On-time delivery %,' 'In-full delivery %,' and 'OnTime in full %' against set targets for daily customer orders to address these issues effectively.

### Github link : https://github.com/shashankhiwarkar/GDS-Mart-s---Supply-Chain-Efficiency-Analysis.git 

### Impoting Data

In [29]:
# importing librabry

import numphy as np
import pandas as pd

In [100]:
# Data Loading, reading CSV file

customers=pd.read_csv('Raw Data/dim_customers.csv')

dates=pd.read_csv('Raw Data/dim_date.csv')

products=pd.read_csv('Raw Data/dim_products.csv')

target_orders=pd.read_csv('Raw Data/dim_targets_orders.csv')

order_line=pd.read_csv('Raw Data/fact_order_lines.csv')

order_aggregate=pd.read_csv('Raw Data/fact_orders_aggregate.csv')

### Getting familer with Data

### Table : fact_order_line

In [30]:
#stored order_line data  to DataFrame stored as orderline

orderline=pd.DataFrame(order_line)
orderline

order_id order_placement_date  customer_id  product_id  order_qty  \
0        FMR32603203            01-Mar-22       789603    25891203        460   
1        FMR33401603            01-Mar-22       789401    25891203        429   
2        FMR32320302            01-Mar-22       789320    25891203        347   
3        FMR33320501            01-Mar-22       789320    25891203        187   
4        FMR34220601            01-Mar-22       789220    25891203        235   
...              ...                  ...          ...         ...        ...   
57091  FAUG831420303            30-Aug-22       789420    25891303         24   
57092  FAUG831122303            30-Aug-22       789122    25891303         67   
57093  FAUG831702303            30-Aug-22       789702    25891303         41   
57094  FAUG831221503            30-Aug-22       789221    25891303         92   
57095  FAUG831202303            30-Aug-22       789202    25891303         29   

      agreed_delivery_date actual_delivery_date  delivery_qty  In Full  \
0                02-Mar-22            03-Mar-22           460        1   
1                03-Mar-22            03-Mar-22           429        1   
2                02-Mar-22            02-Mar-22           347        1   
3                03-Mar-22            03-Mar-22           150        0   
4                04-Mar-22            04-Mar-22           235        1   
...                    ...                  ...           ...      ...   
57091            31-Aug-22            30-Aug-22            24        1   
57092            31-Aug-22            31-Aug-22            54        0   
57093            31-Aug-22            30-Aug-22            37        0   
57094            31-Aug-22            31-Aug-22            83        0   
57095            31-Aug-22            01-Sep-22            29        1   

       On Time  On Time In Full  
0            0                1  
1            1                1  
2            1                1  
3            1                0  
4            1                1  
...        ...              ...  
57091        1                1  
57092        1                0  
57093        1                0  
57094        1                0  
57095        0                1  

[57096 rows x 11 columns]

In [37]:
#Understanding rows and cloumns of data using shape of the data

orderline.shape

(57096, 11)

In [31]:
## check first 10 records using head

orderline.head(10)

order_id order_placement_date  customer_id  product_id  order_qty  \
0  FMR32603203            01-Mar-22       789603    25891203        460   
1  FMR33401603            01-Mar-22       789401    25891203        429   
2  FMR32320302            01-Mar-22       789320    25891203        347   
3  FMR33320501            01-Mar-22       789320    25891203        187   
4  FMR34220601            01-Mar-22       789220    25891203        235   
5  FMR33703603            01-Mar-22       789703    25891203        176   
6  FMR33721603            01-Mar-22       789721    25891203        345   
7  FMR33420203            01-Mar-22       789420    25891203        138   
8  FMR34420402            01-Mar-22       789420    25891203        381   
9  FMR32403401            01-Mar-22       789403    25891203        348   

  agreed_delivery_date actual_delivery_date  delivery_qty  In Full  On Time  \
0            02-Mar-22            03-Mar-22           460        1        0   
1            03-Mar-22            03-Mar-22           429        1        1   
2            02-Mar-22            02-Mar-22           347        1        1   
3            03-Mar-22            03-Mar-22           150        0        1   
4            04-Mar-22            04-Mar-22           235        1        1   
5            03-Mar-22            03-Mar-22           176        1        1   
6            03-Mar-22            03-Mar-22           345        1        1   
7            03-Mar-22            06-Mar-22           138        1        0   
8            04-Mar-22            05-Mar-22           381        1        0   
9            02-Mar-22            02-Mar-22           348        1        1   

   On Time In Full  
0                1  
1                1  
2                1  
3                0  
4                1  
5                1  
6                1  
7                1  
8                1  
9                1

In [32]:
# understanding the totol column 
orderline.columns

Index(['order_id', 'order_placement_date', 'customer_id', 'product_id',
       'order_qty', 'agreed_delivery_date', 'actual_delivery_date',
       'delivery_qty', 'In Full', 'On Time', 'On Time In Full'],
      dtype='object')

In [35]:
# Checking null values
orderline.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 57096 entries, 0 to 57095
Data columns (total 11 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   order_id              57096 non-null  object
 1   order_placement_date  57096 non-null  object
 2   customer_id           57096 non-null  int64 
 3   product_id            57096 non-null  int64 
 4   order_qty             57096 non-null  int64 
 5   agreed_delivery_date  57096 non-null  object
 6   actual_delivery_date  57096 non-null  object
 7   delivery_qty          57096 non-null  int64 
 8   In Full               57096 non-null  int64 
 9   On Time               57096 non-null  int64 
 10  On Time In Full       57096 non-null  int64 
dtypes: int64(7), object(4)
memory usage: 4.8+ MB


### Metric 1 : Total Order Lines

In [89]:
total_order_lines=len(orderline['order_id'])
print("Total Order_line count is:",total_order_lines)

Total Order_line count is: 57096


#### There have been a total of 57,096 order lines processed, indicating a significant volume of orders.

### Metric 2 : Line Fill Rate

In [61]:
x = []
for i in orderline['In Full']:
    x.append(i)
num_order_line  = sum(x)


line_fill_rate=round(num_order_line/total_order_lines,2)
line_fill_rate
print("LIFR%:",line_fill_rate)

LIFR%: 0.66


#### The Line Fill Rate, which is at 0.66 or 66%, suggests that approximately 66% of ordered items were delivered in full. This metric assesses the accuracy of order fulfillment.

### Metric 3 : Volume Fill rate

In [62]:
rate  = sum(orderline['delivery_qty']) / sum(orderline['order_qty'])
VOFR=round(rate,2)
print("VOFR%:",VOFR)

VOFR%: 0.97


#### The Volume Fill Rate, at 0.97 or 97%, indicates that approximately 97% of the ordered quantities were delivered. This metric measures the quantity of products delivered in relation to what was ordered.

### Table : fact_orders_aggregate

In [63]:
# creating dataframe
orderaggregate=pd.DataFrame(order_aggregate)

orderaggregate

order_id  customer_id order_placement_date  on_time  in_full  otif
0        FMR32103503       789103            01-Mar-22        1        0     0
1        FMR34103403       789103            01-Mar-22        1        0     0
2        FMR32103602       789103            01-Mar-22        1        0     0
3        FMR33103602       789103            01-Mar-22        1        0     0
4        FMR33103401       789103            01-Mar-22        1        0     0
...              ...          ...                  ...      ...      ...   ...
31724  FAUG831121302       789121            30-Aug-22        1        1     1
31725  FAUG831421102       789421            30-Aug-22        1        1     1
31726  FAUG831122203       789122            30-Aug-22        1        1     1
31727  FAUG831303403       789303            30-Aug-22        1        1     1
31728  FAUG831703602       789703            30-Aug-22        1        1     1

[31729 rows x 6 columns]

In [64]:
## checking if any null values

orderaggregate.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31729 entries, 0 to 31728
Data columns (total 6 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   order_id              31729 non-null  object
 1   customer_id           31729 non-null  int64 
 2   order_placement_date  31729 non-null  object
 3   on_time               31729 non-null  int64 
 4   in_full               31729 non-null  int64 
 5   otif                  31729 non-null  int64 
dtypes: int64(4), object(2)
memory usage: 1.5+ MB


In [71]:
# checking total rows and count of columns

orderaggregate.shape

(31729, 6)

In [77]:
# checking all the column in list 
orderaggregate.columns

Index(['order_id', 'customer_id', 'order_placement_date', 'on_time', 'in_full',
       'otif'],
      dtype='object')

### Metric 4 : Total Orders

In [70]:
# calculting a total column lenght using len

totalorders=len(orderaggregate['order_id'])
print("total orders:",totalorders)

total orders: 31729


#### There have been 31,729 total orders, showing a high demand for products or services.

### Metric 5:  On time Delivery rate

In [76]:
# calculating ontime delivary

num_order=sum(orderaggregate['on_time'])
on_time_delivery=num_order/totalorders
print("On time Delivery % is",round(on_time_delivery,2))

On time Delivery % is 0.59


#### The On-Time Delivery Percentage is 0.59 or 59%, meaning that 59% of orders were delivered on time as per the agreed-upon schedule.

### Metric 6 : In Full rate

In [85]:
full_order=sum(orderaggregate['in_full'])
in_full_delivery=full_order/totalorders
print("In Full Delivey %",round(in_full_delivery,2))

In Full Delivey % 0.53


#### The In-Full Delivery Percentage is 0.53 or 53%, indicating that 53% of orders were delivered with all items in full. This metric assesses whether orders are complete or if items are missing upon delivery.

### Metric 7 : On Time In full rate

In [88]:
otif=sum(orderaggregate['otif'])
on_time_in_full=otif/totalorders
on_time_in_full
print("On Time In Full %:",round(on_time_in_full,2))

On Time In Full %: 0.29


#### The On-Time In Full percentage is 29%, which combines both on-time delivery and in full delivery performance. This means that 29% of orders are meeting both delivery timeframes and completeness requirements.

### Table : dim_targets_orders

In [90]:
# creating data frame
targetorders=pd.DataFrame(target_orders)
targetorders

customer_id  ontime_target%  infull_target%  otif_target%
0        789201              87              81            70
1        789202              85              81            69
2        789203              92              76            70
3        789301              89              78            69
4        789303              88              78            69
5        789101              86              80            69
6        789102              90              81            73
7        789103              92              67            62
8        789121              78              77            60
9        789122              76              65            49
10       789220              91              76            69
11       789221              85              80            68
12       789320              91              81            74
13       789321              87              81            70
14       789401              85              79            67
15       789402              89              80            71
16       789403              86              79            68
17       789420              75              79            59
18       789421              78              66            51
19       789422              79              81            64
20       789501              92              80            74
21       789503              86              82            71
22       789520              75              68            51
23       789521              78              79            62
24       789522              76              79            60
25       789601              92              67            62
26       789603              92              75            69
27       789702              89              66            59
28       789703              85              78            66
29       789621              92              81            75
30       789622              90              79            71
31       789720              86              77            66
32       789721              89              75            67
33       789902              90              81            73
34       789903              92              65            60

### Metric 8 : Average of on-time target%

In [96]:
avg_on_time_order = round(targetorders['ontime_target%'].mean(),2)
print("Average of on-time target%:",avg_on_time_order)

Average of on-time target%: 86.09


#### The average On-Time Target percentage is 86.09%, suggesting that, on average, orders are achieving 86.09% of their on-time delivery targets.

### Metric 9 : Average of In-Full target

In [99]:
avg_in_full_order = round(targetorders['infull_target%'].mean(),2)
print("Average of In-Full target:",avg_in_full_order)

Average of In-Full target: 76.51


#### The average In-Full Target percentage is 76.51%, indicating that orders are, on average, meeting 76.51% of their completeness targets.

### Average of On Time In Full Delivary

In [98]:
avg_on__time_in_ful_order = round(targetorders['otif_target%'].mean(),2)
print("Average of OTIF:",avg_on__time_in_ful_order)

Average of OTIF: 65.91


#### The average OTIF percentage is 65.91%, which combines both on-time and in full performance. This indicates that, on average, orders are achieving 65.91% of their combined on-time and completeness targets

### Conclusion- To improve, focus on enhancing on-time and in-full delivery percentages for better customer satisfaction and strive to bring the average OTIF percentage closer to 100% for comprehensive order fulfillment efficiency.